sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -156,6 +156,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val ALLOW_NESTEDJOIN_FALLBACK = buildConf("spark.sql.join.broadcastJoinFallback.enabled")
+    .internal()
+    .doc("When true (default), if the other options are not available, fallback to try and use " +
+      "BroadcastNestedLoopJoin as join strategy. This can cause OOM which can be a problem " +
+      "in some scenarios, eg. when running the thriftserver. Turn to false to disable it: an " +
+      "AnalysisException will be thrown.")
```

Comment:
I do not think this can resolve the issue. The fix is in a wrong direction.

Reply:
why? May you please explain which would be the right direction then?

Reply:
OOM could be triggered by various cases. `BroadcastNestedLoopJoin` is being widely used in many cases. If this is the case in your scenario, you can do it in your fork.

Also, this PR just helps a very specific case. I do not think it resolves the general issue.

Reply:
I see and in general I agree with you, but in this case `BroadcastNestedLoopJoin` is used as a fallback even though we are over the threshold for the broadcast joins. So it is very likely that this is going to throw an OOM.
So the rationale of this choice seems to me: I don't have any working join implementation, then just go for this one even though I know in advance that a OOM is likely to occur; anyway I don't have other choices, so if the OOM won't occur I have been lucky and I have been able to run it, otherwise it throws an OOM instead of an `AnalysisException` (so I am not able to run this scenario anyway). And this is fine if we think to normal Spark application, while for long running ones like STS an OOM is much worse than an `AnalysisException`.
Therefore, I know that this is not the solution to all OOM problem, but the reason of this PR is that I think that when this rule was introduced, the choice was made "forgetting" of use cases like STS. Do you agree with me?

Reply:
`AUTO_BROADCASTJOIN_THRESHOLD` is not the threshold for avoiding OOM. It is normally used for performance tuning. 

`BroadcastNestedLoopJoin ` is used for most cases when we do not have the equi join keys. Thus, disabling this join algorithm makes Spark SQL unable to handle many join cases. Thus, I do not think it makes sense to add this conf

Reply:
@gatorsmile I think it's not just solve specific case, this is general in our production environment, users use sparksql and the might not so clear about the size of table, I think it does make sense to provide an option for disable BroadcastNestedLoopJoin for large dataset other than let it go which may cause OOM and make other user's query aborted.

I think this PR should also consider the sizeInBytes of LogicalPlan